File: 2015/deno/day05/part-1.ts
export function processPart1(input: string): number {
	const trimmed = input.trim();
	const lines = trimmed.split('\n');

	let niceStrings = 0;

	for (const line of lines) {
		if (isNice(line)) {
			niceStrings++;
		}
	}
	return niceStrings;
}

function isNice(s: string): boolean {
	if (!threeVowels(s)) {
		return false;
	}
	if (!doubleLetters(s)) {
		return false;
	}
	if (badStrings(s)) {
		return false;
	}
	return true;
}

function threeVowels(s: string): boolean {
	const vowels = 'aeiou';
	let vowelCount = 0;

	for (const char of s.split('')) {
		if (vowels.includes(char)) {
			vowelCount++;
		}
	}
	if (vowelCount >= 3) {
		return true;
	}
	return false;
}

function doubleLetters(s: string): boolean {
	for (let i = 1; i < s.length; i++) {
		if (s[i] === s[i - 1]) {
			return true;
		}
	}
	return false;
}

function badStrings(s: string): boolean {
	const badStrings = ['ab', 'cd', 'pq', 'xy'];

	for (const badString of badStrings) {
		if (s.includes(badString)) {
			return true;
		}
	}
	return false;
}

File: 2015/go/day05/part1.go
package main

import "strings"

func processPart1(input string) int {
	lines := strings.Split(input, "\n")
	niceStrings := 0

	for _, line := range lines {
		if isNice(line) {
			niceStrings++
		}

	}

	return niceStrings
}

func threeVowels(s string) bool {
	vowels := "aeiou"
	count := 0

	for _, char := range s {
		if strings.ContainsRune(vowels, char) {
			count++
		}
		if count >= 3 {
			return true
		}
	}

	return false
}

func doubles(s string) bool {
	for i := 1; i < len(s); i++ {
		if s[i] == s[i-1] {
			return true
		}
	}

	return false
}

func badStrings(s string) bool {
	badStrings := []string{"ab", "cd", "pq", "xy"}

	for _, str := range badStrings {
		if strings.Contains(s, str) {
			return true
		}
	}
	return false
}

func isNice(s string) bool {
	if !threeVowels(s) {
		return false
	}
	if !doubles(s) {
		return false
	}
	if badStrings(s) {
		return false
	}
	return true
}
